java/rocksjni/type_util.cc

```diff
@@ -0,0 +1,275 @@
+    jint user_key_len, jbyteArray int_key, jint int_key_off, jint int_key_len,
+    jlong options_handle) {
+  auto *options =
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  std::string seek_key_buf;
+  auto getInternalKeySeek = [&seek_key_buf,
+                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
+    ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
+        target_slice, options->comparator, &seek_key_buf);
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
+                                          user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
+  return ROCKSDB_NAMESPACE::JniUtil::copyToIndirect(env, key_slice, int_key,
+                                                    int_key_off, int_key_len);
+}
+
+/*
+ * Class:     org_rocksdb_TypeUtil
+ * Method:    getInternalKeyByteArrayForPrev1
+ * Signature: ([BII[BIIJ)I
+ */
+jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyByteArrayForPrev1(
+    JNIEnv *env, jclass /*cls*/, jbyteArray user_key, jint user_key_off,
+    jint user_key_len, jbyteArray int_key, jint int_key_off, jint int_key_len,
+    jlong options_handle) {
+  auto *options =
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  std::string seek_key_buf;
+  auto getInternalKeySeek = [&seek_key_buf,
+                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
+    ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
+        target_slice, options->comparator, &seek_key_buf);
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
+                                            user_key_off, user_key_len);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
+  return ROCKSDB_NAMESPACE::JniUtil::copyToIndirect(env, key_slice, int_key,
+                                                    int_key_off, int_key_len);
+}
+
+/*
+ * Class:     org_rocksdb_TypeUtil
+ * Method:    getInternalKeyForPrevJni
+ * Signature: ([BIJ)[B
+ */
+jbyteArray JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyForPrevJni(
+    JNIEnv *env, jclass /*cls*/, jbyteArray user_key, jint user_key_len,
+    jlong options_handle) {
+  jbyte *target = env->GetByteArrayElements(user_key, nullptr);
+  if (target == nullptr) {
+    // exception thrown: OutOfMemoryError
+    return nullptr;
+  }
+  auto *options =
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  ROCKSDB_NAMESPACE::Slice target_slice(reinterpret_cast<char *>(target),
+                                        user_key_len);
+  std::string seek_key_buf;
+  ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
+      target_slice, options->comparator, &seek_key_buf);
+  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
+  jbyteArray jkey = env->NewByteArray(static_cast<jsize>(key_slice.size()));
+  if (jkey == nullptr) {
```

Comment:
Please use the utility exception class in portal.h

Reply:
done